Require sync "when" invocations after resolution

kelunik opened the pull request:
It specifies the exact invocation parameter values now and includes a statement about always-sync when callbacks.

Closes #20.

kelunik committed:
Clarify Promise::when docblock

It specifies the exact invocation parameter values now and includes a statement about always-sync when callbacks.

Closes #20.

joshdifabio commented:
Does this mean the following needs to be removed from the spec? 

> All registered callbacks MUST be executed in the order they were registered.

I'm thinking about what should happen if a promise has just been resolved and is in the process of firing it's callbacks (or has deferred doing so) when `when()` is called.

bwoebi commented:
@joshdifabio no, that should still take priority.

EDIT: looks like I was misunderstanding you…

joshdifabio commented:
Even though amp doesn't currently satisfy this requirement? 

> All registered callbacks MUST be executed in the order they were registered.

Is it clear what I mean?

joshdifabio commented:
Here's a code example.

```php
function ($promise) {
    $promise->when(function () use ($promise) {
        echo '1';
        $promise->when(function () {
            echo '3';
        });
    });

    $promise->when(function () {
        echo '2';
    });

    $promise->resolve('foo');
}
```

I think this would print _132_ in Amp, even though according to the spec it should print _123_? Always-sync `when()` on resolved promises makes it quite messy to avoid this.

kelunik commented:
@joshdifabio Amp will soon be fixed, I wasn't aware of that.

kelunik commented:
On the other hand, we might just drop that requirement, dunno whether it's important.

joshdifabio commented:
I think the existing requirement regarding the invocation order of callbacks is much more important than the one added in this PR, and while it's probably not that hard to support both requirements in Amp it would be extremely difficult for libs with task queues which are deferring invocation to do so.

kelunik commented:
@joshdifabio If we merge this PR, then there MUST NOT be any task queues. This PR exist to eliminate them.

joshdifabio commented:
> @joshdifabio If we merge this PR, then there MUST NOT be any task queues. This PR exist to eliminate them.

Aha! I hadn't realised.

kelunik commented:
I just made the title of this PR clear. It does way more than it actually should. The clarification should be on a separate PR.

kelunik commented:
@jsor Are you fine with that? 10 hours ago you said you'd prefer unspecified, but then you said a always sync implementation is possible in React.

joshdifabio commented:
> @joshdifabio If we merge this PR, then there MUST NOT be any task queues. This PR exist to eliminate them.

I just want to clarify; are you saying that React and others would have to completely remove their task queues? (I believe that they would, unless we stop requiring that callbacks are fired in a particular order.)

kelunik commented:
> I just want to clarify; are you saying that React and others would have to completely remove their task queues? 

They can do what they want. Just not for `when`. They can continue with task queues for `then`.

bwoebi commented:
I'm not sure, but due to this being a little ambiguous in the text, could you please just specify a bit more precisely when what is called (immediate vs. queue).

joshdifabio commented:
As per amphp/amp#57, the following two requirements are mutually exclusive:

1. If the promise is already resolved, the callback MUST be executed immediately.
2. All registered callbacks MUST be executed in the order they were registered.

So this PR needs to also remove or modify the second clause.

kelunik committed:
Change Promise callback invocation order

Callbacks can't be executed always-sync and in the same order as registered in the edge case where another when handler is registered within a when handler. Once a promise is resolved, all new when handlers must be executed immediately now.

kelunik commented:
I just adjusted the PR.

bwoebi commented:
Now it's unambiguous :-)

joshdifabio commented:
Looks good. Thanks @kelunik!

kelunik commented:
I'd like to await @jsor's approval for this PR.

jsor commented:
@kelunik :shipit:
